### ISSUE TITLE
SP timeout error at when processing slowly

### ISSUE THREAD

**Issue by pgg123** [waveform80/picamera#169]
I have some picamera 1.8 code that seems to work pretty well. However, if I add a little sleep of about 1/2 a sec, I get the SP timeout error.
This code is just the motion detection, record with pre motion and HTTP streaming examples. 
The code below works. To get it to fail, uncomment the sleep at line 145 (labelled This causes the SP timeout error). You need to access the server on port 8080 and move your hand in front of the camera to get all the code to run simultaneously.

```python
import cv2
import datetime
import io
import picamera
import time
import threading
import numpy as np

from BaseHTTPServer import BaseHTTPRequestHandler,HTTPServer

VIDEO = 1
HTTP = 2
MOTION = 0


# A simple Motion Detection algorithm.
class MotionDetection:
    def __init__(self, camera):
        self.wipe_images()

        self._MOTION_LEVEL = 200000
        self._THRESHOLD = 35

    def wipe_images(self):
        self.__image0 = None
        self.__image1 = None
        self.__image2 = None

    def get_image(self):
        stream = io.BytesIO()
        camera.capture(stream, format='jpeg', use_video_port=True, splitter_port=MOTION)
        data = np.fromstring(stream.getvalue(), dtype=np.uint8)
        return cv2.imdecode(data, 1)

    def _update_image(self):
        self.__image2 = self.__image1 if self.__image1 is not None else self.get_image()
        self.__image1 = self.__image0 if self.__image0 is not None else self.get_image()
        self.__image0 = self.get_image()

    def _get_motion(self):
        d1 = cv2.absdiff(self.__image1, self.__image0)
        d2 = cv2.absdiff(self.__image2, self.__image0)
        result = cv2.bitwise_and(d1, d2)

        (value, result) = cv2.threshold(
            result, self._THRESHOLD, 255, cv2.THRESH_BINARY)

        scalar = cv2.sumElems(result)
        return scalar

    def detect_motion(self):
        self._update_image()
        motion = self._get_motion()
        if motion and motion[0] > self._MOTION_LEVEL:
            self.wipe_images()
            return True
        return False

    def save_current_image(self, file_name):
        cv2.imwrite(file_name, self.__image0)


def save_before_motion_video(stream, fname):
    # Write the entire content of the circular buffer to disk
    with io.open(fname, 'wb') as output:
        with stream.lock:
            for frame in stream.frames:
                if frame.header:
                    stream.seek(frame.position)
                    break
            while True:
                buf = stream.read1()
                if not buf:
                    break
                output.write(buf)


def mainx(camera):
    stream = picamera.PiCameraCircularIO(camera, seconds=2)
    camera.start_recording(stream, format='h264', splitter_port=VIDEO)

    detection = MotionDetection(camera)
    count = 1

    print "READY"

    try:
        while True:

            camera.wait_recording(1, splitter_port=VIDEO)

            print "Looking......"
            if detection.detect_motion():
                print "motion has been detected"

                # As soon as we detect motion, split the recording to record
                # the bits "after" motion
                tstmp = datetime.datetime.now().strftime("%Y-%m-%d-%H%M%S")
                fname  = "motion_%s_%05d.h264" % (tstmp, count)
                print "motion file name: ", fname
                count += 1
                camera.split_recording("data/" + fname, splitter_port=VIDEO)

                # While that's going, write the 10 seconds "before" motion to
                # disk as well, and wipe the circular stream
                save_before_motion_video(stream, "data/pre_" + fname)

                # Wait for trigger period, then split recording
                # back to the in-memory circular buffer
                for _ in range(5):
                    print "sleep and poll"
                    camera.wait_recording(1, splitter_port=VIDEO)

                print "about to split recording"
                camera.split_recording(stream, splitter_port=VIDEO)
    except Exception as e:
        print "EXCEPTION HERE", str(e)
    finally:
        print "b4 stop_recoding"
        camera.stop_recording(splitter_port=VIDEO)




class CamHandler(BaseHTTPRequestHandler):
  def do_GET(self):
    if self.path.endswith('.mjpg'):
      self.send_response(200)
      self.send_header('Content-type','multipart/x-mixed-replace; boundary=--jpgboundary')
      self.end_headers()
      stream=io.BytesIO()
      try:
        start=time.time()
        for foo in camera.capture_continuous(stream,'jpeg',use_video_port=True, splitter_port=HTTP):
          print "IMAGE JUST TAKEN FOR HTTP"
          self.wfile.write("--jpgboundary")
          self.send_header('Content-type','image/jpeg')
          self.send_header('Content-length',len(stream.getvalue()))
          self.end_headers()
          self.wfile.write(stream.getvalue())
          stream.seek(0)
          stream.truncate()
###### This causes the SP timeout error          time.sleep(.5)
### Timed out waiting for an SPS header
          time.sleep(0.1) # This is OK
      except KeyboardInterrupt:
        print "EXCEPTION"*20
      return
    else:
      self.send_response(200)
      self.send_header('Content-type','text/html')
      self.end_headers()
      self.wfile.write("""<html><head></head><body>
        <img src="/cam.mjpg"/>
      </body></html>""")
      return

def main():
  global camera
  camera = picamera.PiCamera()
  camera.resolution = (640, 480)
  global img
  try:
    server = HTTPServer(('',8080), CamHandler)
    print "server started"
    threading.Thread(target=server.serve_forever).start()
    time.sleep(2)
  except KeyboardInterrupt:
    print "EXCEPTION 1 "*10
    camera.close()
    server.socket.close()
  mainx(camera)
  print "00000000 ERROR 000000"

if __name__ == '__main__':
  main()
```

**Comment by waveform80** [comment]
Okay, played with this for a bit and I'm still not sure what's going on. I thought this might be a GIL/concurrency thing but that can't be it (as it's a problem with *increasing* a sleep time, making a thread do less work rather than the opposite). It might be something odd with using splits with `capture_continuous` but I'm not sure whether that's a bug in picamera or simply a use-case to be avoided.

For now, I'm going to try and come up with a simpler test-case to try and figure out the real issue here but given 1.9's got several rather important bug fixes in it (like memory leaks) I'm reluctantly going to push this one to 1.10 so it doesn't hold up the 1.9 release.

**Comment by waveform80** [comment]
Sorry, this is going to have to be pushed to 1.11 - I need to get 1.10 out quickly as I'm going to be spending time on family obligations for a while!

**Comment by waveform80** [comment]
Thanks for the full reproduction code. I've tried this with the nearly-finished 1.11 and the it now seems to work happily even with the marked line uncommented (I'm fairly sure I tried this a while ago and it still failed). If I had to guess, I'd say 8c505686e726739df3070ed5302e9aea0fad0f28 probably fixes it (inadvertently). It's a bit of a hack as fixes go, but the most reasonable I can manage without significant amounts of experimentation.
